### DISCUSSION TITLE
GCP Cloud Run - Distributed Cache

### DISCUSSION THREAD

**Discussion by besserwisser** [keycloak/keycloak#23458] [Q&A]
Hello community,

Google Cloud Run is one of the main services promoted by Google. It is also easy to set up Keycloak with it (container Image) and and external database. But one thing I don't understand is, how you would handle the cache, when you have more then one instance due to auto scaling.

Can you have a shared Keycloak cache in Cloud Run? I read about the GOOGLE_PING2 discovery protocol with Infinispan. But does this apply to the Cloud Run Service? 

Thank you!

**Comment by clintonb** [comment]
`KC_CACHE_STACK=kubernetes` fails:

```
{
  "jsonPayload": {
    "mdc": {},
    "threadId": 1,
    "sequence": 10050,
    "loggerName": "org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler",
    "hostName": "localhost",
    "ndc": "",
    "message": "ERROR: org.infinispan.manager.EmbeddedCacheManagerStartupException: ISPN000541: Error while trying to create a channel using the specified configuration '[TCP(bundler.max_size=64000, sock_conn_timeout=300, linger=-1, port_range=0, thread_pool.keep_alive_time=60000, diag.enabled=false, bind_port=7800, thread_naming_pattern=pl, thread_pool.thread_dumps_threshold=10000, send_buf_size=640k, thread_pool.max_threads=200, use_virtual_threads=false, bundler_type=transfer-queue, logical_addr_cache_expiration=360000, thread_pool.min_threads=0), RED(), dns.DNS_PING(num_discovery_runs=3), MERGE3(max_interval=30000, min_interval=10000), FD_SOCK2(offset=50000), FD_ALL3(), VERIFY_SUSPECT2(timeout=1000), pbcast.NAKACK2(xmit_table_num_rows=50, use_mcast_xmit=false, xmit_table_msgs_per_row=1024, xmit_table_max_compaction_time=30000, xmit_interval=200, resend_last_seqno=true), UNICAST3(conn_close_timeout=5000, xmit_interval=200, xmit_table_num_rows=50, xmit_table_msgs_per_row=1024, xmit_table_max_compaction_time=30000), pbcast.STABLE(desired_avg_gossip=5000, max_bytes=1M), pbcast.GMS(join_timeout=2000, print_local_addr=false), UFC(min_threshold=0.40, max_credits=4m), MFC(min_threshold=0.40, max_credits=4m), FRAG4(frag_size=60000)]'",
    "processName": "QuarkusEntryPoint",
    "level": "ERROR",
    "timestamp": "2024-04-07T05:44:47.241765679Z",
    "loggerClassName": "org.jboss.logging.Logger",
    "threadName": "main",
    "processId": 1
  }
}
```

Using `google` also fails with this message:

```
ERROR: JGRP000002: unable to load protocol org.jgroups.protocols.google.GOOGLE_PING2 (either with relative - org.jgroups.protocols.google.GOOGLE_PING2 - or absolute - org.jgroups.protocols.org.jgroups.protocols.google.GOOGLE_PING2 - class name)
```

It doesn't seem distributed caching is supported on Cloud Run.

**Comment by paulwer** [comment]
@besserwisser @clintonb had some of you success with this topic or found an alternative solution?

**Reply by clintonb** [reply]
Nope. I’m actively moving away from Cloud Run due to recent incidents and issues with auto scaling (when using a sidecar).

**Comment by Jojoooo1** [comment]
Also interested!

**Comment by lucas-giaco** [comment]
If anyone is interested, I was able to make `GOOGLE_PING2` work with CloudRun.

As stated in the [docs](https://www.keycloak.org/server/caching#_additional_transport_stacks), you need to add the required JARs to your Keycloak image for cloud vendor-specific stacks.

Here's a list of the JARs I had to add to make it work:
* api-common-1.9.2.jar
* failureaccess-1.0.1.jar
* gax-1.57.0.jar
* gax-httpjson-0.74.0.jar
* google-api-client-1.30.9.jar
* google-api-services-storage-v1-rev20200430-1.30.9.jar
* google-auth-library-credentials-0.20.0.jar
* google-auth-library-oauth2-http-0.20.0.jar
* google-cloud-core-1.93.6.jar
* google-cloud-core-http-1.93.6.jar
* google-cloud-storage-1.111.0.jar
* google-http-client-1.35.0.jar
* google-http-client-appengine-1.35.0.jar
* google-http-client-jackson2-1.35.0.jar
* jgroups-google-2.0.0.Final.jar
* opencensus-api-0.24.0.jar
* opencensus-contrib-http-util-0.24.0.jar
* threetenbp-1.4.4.jar

**Reply by Jojoooo1** [reply]
Amazing, would you mind sharing your configuration ? Did you hit any limitation when running multiple cloud run instances ?

**Reply by lucas-giaco** [reply]
Sure, my Dockerfile is as follows

```dockerfile
ARG KEYCLOAK_VERSION=26.0.6
FROM keycloak/keycloak:${KEYCLOAK_VERSION} AS builder

# Enable health and metrics support
ENV KC_HEALTH_ENABLED=true
ENV KC_METRICS_ENABLED=true

# Configure a database vendor
ENV KC_DB=postgres

WORKDIR /opt/keycloak
COPY providers/ providers/
RUN /opt/keycloak/bin/kc.sh build

FROM keycloak/keycloak:${KEYCLOAK_VERSION}
COPY --from=builder /opt/keycloak/ /opt/keycloak/

ENTRYPOINT ["/opt/keycloak/bin/kc.sh"]
CMD ["start", "--optimized"]
```

To set up `GOOGLE_PING2` you must add 2 configs/env variables, `KC_CACHE_STACK="google"` and `JAVA_OPTS_APPEND="-Djgroups.google.bucket_name=<your-bucket-name>"`

> Did you hit any limitation when running multiple cloud run instances?

So far we haven't, but we haven't run any load tests neither

**Reply by paulwer** [reply]
Dont know if thats the right place, but what prices can be estimated by using Google_Ping in gcr?p
I havnt worked with it yet and did not find an article or resourced about this service related to pricing.
